rrs/tools/upgrade_history_internal.py

```diff
@@ -0,0 +1,617 @@
+            if not f in ct_files:
+                ct_files.append(f)
+
+    # Check moves for recipe -> inc with an added recipe
+    # (i.e. the move should really be to the newly added recipe)
+    # example: d5a95dc8985a42bb7e50bc4e7dc6b012d711ff08 in OE-Core
+    for i,(a,b) in enumerate(moved_files):
+        if b.endswith('.inc'):
+            for af in added_files:
+                # This is naive, but good enough
+                if af.rsplit('_')[0] == a.rsplit('_')[0]:
+                    moved_files[i] = (a,af)
+
+    return ct_files, deleted, moved_files
+
+
+def checkout_layer_deps(layerbranch, commit, fetchdir, logger):
+    """ Check out the repositories for a layer and its dependencies """
+
+    oecore_map = {}
+    if layerbranch.layer.name != 'openembedded-core':
+        # Filter out some bad commits for OE-Core
+        for good_rev, bad_revs in oecore_bad_revs.items():
+            for bad_rev in bad_revs:
+                oecore_map[bad_rev] = good_rev
+
+    # Some layers will be in the same repository, so we only want to check those out once
+    done_repos = []
+    def checkout_layer(lb, lcommit=None, lcommitdate=None, force=False):
+        urldir = str(lb.layer.get_fetch_dir())
+        repodir = os.path.join(fetchdir, urldir)
+        if not repodir in done_repos:
+            if not lcommit:
+                lcommit = utils.runcmd(['git', 'rev-list', '-1', '--before=%s' % lcommitdate, 'origin/master'], repodir, logger=logger).strip()
+            if lb.layer.name == 'openembedded-core':
+                lmapcommit = oecore_map.get(lcommit, None)
+                if lmapcommit:
+                    logger.debug('Preferring OE-Core revision %s over %s' % (lmapcommit, lcommit))
+                    lcommit = lmapcommit
+            utils.checkout_repo(repodir, lcommit, logger, force)
+            if lcommit in oecore_bad_revs_2:
+                # Fix issue that was introduced in 309a02931779f32d1139cc1169a039cbe4638706 and fixed in 40a904bf8bc1279c3da0893c003f740f1d2066c2
+                with open(os.path.join(repodir, 'meta/conf/bitbake.conf'), 'r') as f:
+                    lines = f.readlines()
+                lines.insert(0, 'BBINCLUDED ?= ""\n')
+                with open(os.path.join(repodir, 'meta/conf/bitbake.conf'), 'w') as f:
+                    f.writelines(lines)
+            done_repos.append(repodir)
+
+    # We "force" here because it's almost certain we'll be checking out a
+    # different revision for the layer itself
+    checkout_layer(layerbranch, commit, force=True)
+    layer_urldir = str(layerbranch.layer.get_fetch_dir())
+    layer_repodir = os.path.join(fetchdir, layer_urldir)
+    commitdate = utils.runcmd(['git', 'show', '-s', '--format=%ci'], layer_repodir, logger=logger)
+
+    for dep in layerbranch.get_recursive_dependencies():
+        checkout_layer(dep, lcommitdate=commitdate)
+
+    return commitdate
+
+
+def generate_history(options, layerbranch_id, commit, logger):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function calls 29 other functions. By comparison, 99% of the functions in the CodeGuru reference dataset use fewer. This indicates the function is highly coupled with other functions. A function that is highly coupled with other functions is difficult to understand and its behavior might change unexpectedly when one of its referenced functions is updated. High coupling could also increase the integration test complexity, maintenance cost and technical debt.

We recommend that you simplify this function or break it into multiple functions. For example, consider extracting the code block on lines 452-459 into a separate function.